src/sinol_make/commands/run/__init__.py

```diff
@@ -360,78 +367,190 @@ def print_view(output_file=None):
+		for solution in used_solutions:
+			if solution in config_expected_scores.keys():
+				expected_scores[solution] = {
+					"expected": {},
+					"points": 0
+				}
+
+				for group in used_groups:
+					if group in config_expected_scores[solution]["expected"]:
+						expected_scores[solution]["expected"][group] = config_expected_scores[solution]["expected"][group]
+
+				expected_scores[solution]["points"] = self.calculate_points(expected_scores[solution]["expected"])
+
+		print(util.bold("Expected scores from config:"))
+		self.print_expected_scores(expected_scores)
+		print(util.bold("\nExpected scores based on results:"))
+		self.print_expected_scores(new_expected_scores)
+
+		expected_scores_diff = dictdiffer.diff(expected_scores, new_expected_scores)
+		added_solutions = set()
+		removed_solutions = set()
+		added_groups = set()
+		removed_groups = set()
+		changes = [] # Array of tuples (solution, group, current_result, expected_result)
+
+		for type, field, change in list(expected_scores_diff):
+			if type == "add":
+				if field == '': # Solutions were added
+					for solution in change:
+						added_solutions.add(solution[0])
+				elif field[1] == "expected": # Groups were added
+					for group in change:
+						added_groups.add(group[0])
+			elif type == "remove":
+				# We check whether a solution was removed only when sinol_make was run on all of them
+				if field == '' and self.args.programs == None and config_expected_scores:
+					for solution in change:
+						removed_solutions.add(solution[0])
+				# We check whether a group was removed only when sinol_make was run on all of them
+				elif field[1] == "expected" and self.args.tests == None and config_expected_scores:
+					for group in change:
+						removed_groups.add(group[0])
+			elif type == "change":
+				if field[1] == "expected": # Results for at least one group has changed
+					solution = field[0]
+					group = field[2]
+					old_result = change[0]
+					result = change[1]
+					changes.append((solution, group, result, old_result))
+
+		return {
+			"added_solutions": added_solutions,
+			"removed_solutions": removed_solutions,
+			"added_groups": added_groups,
+			"removed_groups": removed_groups,
+			"changes": changes,
+			"expected_scores": expected_scores,
+			"new_expected_scores": new_expected_scores,
+		}
+
+
+	def print_expected_scores_diff(self, validation_results):
+		added_solutions = validation_results["added_solutions"]
+		removed_solutions = validation_results["removed_solutions"]
+		added_groups = validation_results["added_groups"]
+		removed_groups = validation_results["removed_groups"]
+		changes = validation_results["changes"]
+		expected_scores = validation_results["expected_scores"]
+		new_expected_scores = validation_results["new_expected_scores"]
+		config_expected_scores = self.config["sinol_expected_scores"] if "sinol_expected_scores" in self.config else {}
```

Comment:
That looks like something that could be a `struct` (or whatever Python has). Having named fields reduces the risk of making a typo in the strings. Then, the function could e.g. take `validation_results` and then abbreviate it to `diff` or whatever, so that the names `diff.added_solutions` won't be long. That would also shorten the code thanks to not saving each dictionary value into a variable.

Reply:
Changed in 189dd7f695d152db99a0ee7d6ddde26c4e1c1a45, 4e272b3503f388c825144586b1570e6b0b986c19 and 8e0951d73e667b2e353d0501fbfdcb2219a98c23.